reactor-netty-http/src/test/java/reactor/netty/http/server/HttpServerTests.java

```diff
@@ -2109,4 +2111,87 @@ else if (msg instanceof ByteBuf) {
 		            .expectComplete()
 		            .verify(Duration.ofSeconds(30));
 	}
+
+	@Test
+	public void testCustomMetricsRecorderWithUriMapper() throws InterruptedException {
+		CountDownLatch latch = new CountDownLatch(1);
+		List<String> collectedUris = new CopyOnWriteArrayList<>();
+
+		disposableServer =
+				HttpServer.create()
+							.port(0)
+							.handle((req, resp) -> resp.sendString(Mono.just("OK")))
+							.wiretap(true)
+							.metrics(true,
+									() -> new HttpServerMetricsRecorder() {
+										@Override
+										public void recordDataReceived(SocketAddress remoteAddress, String uri, long bytes) {
+											collectedUris.add(uri);
+										}
+
+										@Override
+										public void recordDataSent(SocketAddress remoteAddress, String uri, long bytes) {
+											collectedUris.add(uri);
+										}
+
+										@Override
+										public void incrementErrorsCount(SocketAddress remoteAddress, String uri) {
+											collectedUris.add(uri);
+										}
+
+										@Override
+										public void recordDataReceived(SocketAddress remoteAddress, long bytes) {
+										}
+
+										@Override
+										public void recordDataSent(SocketAddress remoteAddress, long bytes) {
+										}
+
+										@Override
+										public void incrementErrorsCount(SocketAddress remoteAddress) {
+										}
+
+										@Override
+										public void recordTlsHandshakeTime(SocketAddress remoteAddress, Duration time, String status) {
+										}
+
+										@Override
+										public void recordConnectTime(SocketAddress remoteAddress, Duration time, String status) {
+										}
+
+										@Override
+										public void recordResolveAddressTime(SocketAddress remoteAddress, Duration time, String status) {
+										}
+
+										@Override
+										public void recordDataReceivedTime(String uri, String method, Duration time) {
+											collectedUris.add(uri);
+										}
+
+										@Override
+										public void recordDataSentTime(String uri, String method, String status, Duration time) {
+											collectedUris.add(uri);
+										}
+
+										@Override
+										public void recordResponseTime(String uri, String method, String status, Duration time) {
+											collectedUris.add(uri);
+										}
+									},
+									s -> s.startsWith("/stream/") ? "/stream/{n}" : s)
+							.bindNow();
+
+		HttpClient.create()
+				.doOnRequest((req, conn) -> conn.onTerminate()
```

Comment:
It is better to count down when every metric is recorded, thus `latch.await` will wait for all to appear.
Wdyt?

Reply:
like initializing as `CountDownLatch(5)` and counting down next to each `collectedUris.add()` call?

why not, especially if you think counting down on the connection termination could lead to making the assertion too early. it has the same drawback as my original concern with `take(5)`, but that is more minor.

I like how the test reads much more with the list 👍

Reply:
yes because terminating is competing with recorder
https://github.com/reactor/reactor-netty/blob/8afc89e173733e97ce67d8bcbf4e3083e8a40f1f/reactor-netty-http/src/main/java/reactor/netty/http/server/HttpTrafficHandler.java#L280
https://github.com/reactor/reactor-netty/blob/8afc89e173733e97ce67d8bcbf4e3083e8a40f1f/reactor-netty-http/src/main/java/reactor/netty/http/server/HttpServerMetricsHandler.java#L77